src/Lecture4.hs

```diff
@@ -134,8 +144,25 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+splitOn :: Char -> String -> [String]
+splitOn delim = uncurry (:) . foldr (\c (container, res) -> if c == delim 
+                  then ([], container : res)
+                  else (c : container, res)
+                ) ([], [])
+
+-- Question: Is there a way to create a PositiveInt type ?
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow = validateValues . splitOn ','
+  where
+    validateValues ["", _, _] = Nothing
+    validateValues [f , s, t] = (readMaybe (dropSpaces s) :: Maybe TradeType) 
+                                  >>= \parsedTradeType -> case readMaybe (dropSpaces t) of
+                                    Just num  -> if num >= 0 
+                                      then Just (Row f parsedTradeType num)
+                                      else Nothing
+                                    Nothing -> Nothing 
+    validateValues _          = Nothing
```

Comment:
You have an excellent monadic implementation 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow str = do
    [f, s, t] <- Just $ splitOn ',' str
    guard $ f /= ""
    parsedTradeType <- readMaybe s
    num <- readMaybe t
    guard $ num >= 0
    pure (Row f parsedTradeType num)
```